hare3/algorithm.go

```diff
@@ -0,0 +1,502 @@
+				Values: []types.Hash20{*p.Li[j]},
+			}
+		} else {
+			for _, c := range candidates {
+				candidateHash := toHash(c.values)
+				// Check to see if valid proposal for this iteration
+				// Round 5 condition c
+				_, ok := p.Ti[j][candidateHash]
+				if !ok {
+					continue
+				}
+				// Check leader
+				// Round 5 condition d
+				if !p.lc.IsLeader(c.id, NewAbsRound(j, 2)) {
+					continue
+				}
+				// Check grade
+				// Round 5 condition e
+				if c.grade < 2 {
+					continue
+				}
+				// Check subset of threshold values at round 3
+				// Round 5 condition f
+				if !isSubset(c.values, p.Vi[2]) {
+					continue
+				}
+				// Check superset of highest graded values or we received a
+				// commit for the previous iteration from thresh-gossip for
+				// this set with grade >= 1.
+				// Round 5 condition g
+				if !isSubset(p.Vi[0], c.values) {
+					// Check for received message
+					lastIterationCommit := NewAbsRound(j-1, 5)
+					values := p.tgg.RetrieveThresholdMessages(lastIterationCommit, 1)
+
+					// If the candidate is not superset of highest graded
+					// values and no commit for that candidate was received in
+					// the previous iteration with grade >= 1 then we go to the
+					// next candidate.
+					if !slices.Contains(values, candidateHash) {
+						continue
+					}
+				}
+				// Locked value for this iteration is nil or matches set
+				//
+				// Round 5 condition h
+				if !(p.Li[j] == nil || *p.Li[j] == candidateHash) {
+					continue
+				}
+				mm = &OutputMessage{
+					Round:  NewAbsRound(j, 5),
+					Values: []types.Hash20{candidateHash},
+				}
+				break
+			}
+		}
+		return mm, nil
+	case Notify:
+		// Case 1
+		values := p.tgg.RetrieveThresholdMessages(NewAbsRound(j-1, 6), 5)
+		resultHash, result := matchInPreviousIterations(values, p.Ti, j)
+		if resultHash != nil {
+			return &OutputMessage{
+				Round:  NewAbsRound(j, 6),
+				Values: []types.Hash20{*resultHash},
+			}, result
+		}
+		// Case 2
+		if active {
+			values := p.tgg.RetrieveThresholdMessages(NewAbsRound(j, 5), 5)
+			setHash, _ := matchInIteration(values, p.Ti, j)
+			return &OutputMessage{
+				Round:  NewAbsRound(j, 6),
+				Values: []types.Hash20{*setHash},
+			}, nil
+
+		}
+	}
+
+	return nil, nil
+}
+
+type OutputMessage struct {
+	Round  AbsRound
+	Values []types.Hash20
+}
+
+// with int8 we have a max iterations of 17 before we overflow.
+// AbsRound (short for absolute round) represents a round of the hare protocol. The hare protocol progresses in iterations and there are 7 rounds per iteration, except for the first iteration which has a one off pre-round  the abs round starts at -1 and increments for each successive round.
+//
+// from the abs round both round and message type can be
+// the iteration and the round within that iteration can be derived.
+type AbsRound int8
+
+// NewAbsRound constructs a new AbsRound from the given iteration (j) and round (r).
+func NewAbsRound(j, r int8) AbsRound {
```

Comment:
i think this type doesn't have an advantage over type where Round and Iteration are stored separately

Reply:
but we will need to fit atleast 40 iterations and max 6 rounds, so 2 bytes either way, right?

we know when to increment iteration and when to increment round, so i thought that it is simpler to have struct with two fields and avoid additional arithmetic

Reply:
>but we will need to fit atleast 40 iterations and max 6 rounds, so 2 bytes either way, right?

Yes 2 bytes either way, by compact I meant we could fit more iterations in if we used a single value, several thousand as opposed to 256, it doesn't seem particularly important to have that though.

>we know when to increment iteration and when to increment round, so i thought that it is simpler to have struct with two fields and avoid additional arithmetic

If you look at the implementation of `NextRound` it just increments the round, it doesn't need to know about iterations or rounds, so we don't really know when to increment iteration or round explicitly in the current implementation.

I feel like if we represented round and iteration separately we'd just end up moving the existing logic to handle updating and calculating the absolute round, but I don't think it would get simpler.

>>{iteration:0, round:7}
>this is invalid, right?

You could consider it invalid or you could consider it as iteration 1 round 0, if you consider it as invalid then that's going to  require some code to validate updates, and that implies errors and that introduces extra complexity.